ios/adapters/epson/NetPrinterEpsonAdapter.m

```diff
@@ -1,112 +1,160 @@
+      long currTime = [[NSDate date] timeIntervalSince1970];
+      while (!_finishedAsyncCall && startTime + delayInSeconds > currTime) {
+        [NSThread sleepForTimeInterval:0.5];
+        currTime = [[NSDate date] timeIntervalSince1970];
+      }
+      if (!_finishedAsyncCall) {
+        [NSException raise:@"Print failed"
+                    format:@"Delegate Function not called"];
+      }
+      [printer endTransaction];
+      [printer disconnect];
+      [printer clearCommandBuffer];
+      [printer setReceiveEventDelegate:nil];
+    } @catch (NSException *exception) {
 
-        _successCallback = successCallback;
-        _errorCallback = errorCallback;
-
-}
-- (void) onPtrReceive:(Epos2Printer *)printerObj code:(int)code status:(Epos2PrinterStatusInfo *)status printJobId:(NSString *)printJobId
-{
-    NSString *errMsg = [EpsonUtils makeErrorMessage:status];
-
-    [printerObj endTransaction];
-    [printerObj clearCommandBuffer];
-    int result = [printerObj disconnect];
-
-    if ([errMsg  isEqual: @""] && result == EPOS2_SUCCESS) {
-        _successCallback != nil ? _successCallback(@[[NSString stringWithFormat:@"Successfuly printed"]]) : nil;
-    } else {
-        _errorCallback != nil ? _errorCallback(@[[NSString stringWithFormat:@"Delegate with ERROR %i and message %@",result,errMsg]]) : nil;
+      [printer endTransaction];
+      int failResult = [printer disconnect];
+      [printer clearCommandBuffer];
+      [printer setReceiveEventDelegate:nil];
+      errorCallback(
+          @[ [NSString stringWithFormat:@"%@ and disconnect code %i",
+                                        exception.reason, failResult] ]);
+    } @finally {
+      _successCallback = nil;
+      _errorCallback = nil;
     }
+  }
+}
 
-    _successCallback = nil;
-    _errorCallback = nil;
-
-    [printerObj setReceiveEventDelegate:nil];
-    return;
+- (void)connectAndSendAux:(NSString *)host
+             printRawData:(NSString *)text
+                  success:(RCTResponseSenderBlock)successCallback
+                     fail:(RCTResponseSenderBlock)errorCallback
+               printerObj:(Epos2Printer *)printer {
+  NSString *target = [NSString stringWithFormat:@"TCP:%@", host];
+  int result = EPOS2_SUCCESS;
+  result = [printer connect:target timeout:5000];
+  if (result != EPOS2_SUCCESS) {
+
+    [NSException
+         raise:@"Invalid connection"
+        format:@"Can't connect to printer %@, ERROR code: %i", host, result];
+  }
+  NSData *payload = [NSData dataWithBase64EncodedString:text];
+  [printer addCommand:payload];
+  result = [printer sendData:5000];
+  if (result != EPOS2_SUCCESS) {
+    [NSException raise:@"Print failed" format:@"Error occurred while printing"];
+  }
+  _successCallback = successCallback;
+  _errorCallback = errorCallback;
+}
+- (void)onPtrReceive:(Epos2Printer *)printerObj
+                code:(int)code
+              status:(Epos2PrinterStatusInfo *)status
+          printJobId:(NSString *)printJobId {
```

Comment:
Nothign changed here, except removal of disconnect code block, the rest is just resolving the ReactNative promised with success or error callback